packages/cli/lib/model-discoverer.js

```diff
@@ -0,0 +1,94 @@
+const debug = require('./debug')('model-discoverer');
+const fs = require('fs');
+const path = require('path');
+
+/**
+ * Given a datasource and discovery options,
+ * return a list of objects {table: 'foo', schema: 'bar}
+ */
+async function discoverModelNames(ds, options) {
+  if (!ds.connected) {
+    await new Promise(resolve => {
+      ds.on('connected', resolve);
+    });
+  }
+  return await ds.discoverModelDefinitions(options);
+}
+
+/**
+ * Returns the entire model definition for a model
+ */
+async function discoverSingleModel(ds, options, name) {
+  const schema = await ds.discoverSchema(name, options);
+  schema.settings = schema.options;
+  return schema;
+}
+
+/**
+ * requires a compiled datasource, instantiates it and returns it
+ * @param path
+ */
+function loadDataSource(path) {
+  const ds = require(path);
+  const key = Object.keys(ds)[0];
+  return new ds[key]();
+}
+
+/**
+ * @param name {string}
+ * @returns {*}
+ */
+function loadDataSourceByName(name) {
+  debug(`Searching for specified dataSource ${name}`);
+  const dataSourceFiles = fs
+    .readdirSync(DEFAULT_DATASOURCE_DIRECTORY)
+    .filter(s => s.endsWith('.datasource.js'));
+  debug(`Loaded ${dataSourceFiles.length} dataSource files`);
+
+  for (let i = 0; i < dataSourceFiles.length; i++) {
+    const f = dataSourceFiles[i];
+    const ds = loadDataSource(path.resolve(DEFAULT_DATASOURCE_DIRECTORY, f));
+    if (ds.name === name) {
+      debug(`Found dataSource ${name}`);
+      return ds;
+    } else {
+      debug(`Did not match dataSource ${name} !== ${ds.name}`);
+    }
+  }
+}
+
+const DEFAULT_DATASOURCE_DIRECTORY = './dist/src/datasources';
+
+const MODEL_TEMPLATE_PATH = path.resolve(
+  __dirname,
+  '../generators/model/templates/model.ts.ejs',
+);
+
+const sanitizeProperty = function(o) {
+  Object.entries(o).forEach(([k, v]) => {
+    // Delete the null properties so the template doesn't spit out `key: ;`
+    if (v === null) {
+      delete o[k];
+    }
+
+    // If you are an object or array, stringify so you don't appear as [object [object]
+    if (v === Object(v)) {
+      o[k] = JSON.stringify(o[k]);
+    }
+  });
+  // o.id could be 1 or 0, typescript will be angry if it is not a boolean
```

Comment:
This is a bug in our type definitions. Can we fix the typedefs to correctly describe `id` as `boolean | number` please?

Reply:
I'm not sure how to do this

Reply:
I think we need to change the following line:

https://github.com/strongloop/loopback-next/blob/6f62444601faba789b798df953d14c2aababf26b/packages/repository/src/model.ts#L31

```diff
-  id?: boolean;
+  id?: boolean | number;
```

Reply:
It may be best to submit this fix as a standalone pull request, that way we can get it landed quickly.

Reply:
I've submitted it here https://github.com/strongloop/loopback-next/pull/2406